### PR TITLE
Separate ConstraintTableDef from IndexTableDef.

### LINKED ISSUE #2430: sql/parser: AST doesn't distinguish between constraints and indexes
In postgres, constraints specifically exclude INDEX, though adding a UNIQUE constraint creates an index. In cockroach, the two appear to be entirely conflated.

### PR DESCRIPTION
**Pull request by petermattis**
Unique constraints (include primary key constraints) are now modelled by
UniqueConstraintTableDef. This lays the groundwork for adding non-index
constraints like a CheckConstraintTableDef.

Fixes #2430.

### CODE REVIEW

File: sql/parser/create.go [hunk 1]
```diff
@@ -182,37 +187,56 @@ type UniqueConstraint struct{}
...
+	constraintTableDef()
+}
+
+func (*UniqueConstraintTableDef) constraintTableDef() {}
+
+// UniqueConstraintTableDef represents a unique constraint within a CREATE
+// TABLE statement.
+type UniqueConstraintTableDef struct {
+	IndexTableDef
+	PrimaryKey bool
```
Comment (petermattis) [inline sql/parser/create.go]: FYI, I originally created a separate `PrimaryKeyConstraintTableDef` struct, but there was huge overlap with that struct and `UniqueConstraintTableDef`.

File: sql/parser/sql.y [hunk 2]
```diff
@@ -1202,13 +1202,22 @@ table_like_option:
...
-      Name:    Name($3),
-      Unique:  $1,
-      Columns: NameList($5),
-      Storing: $7,
+      Name:    Name($2),
+      Columns: NameList($4),
+      Storing: $6,
+    }
+  }
+| UNIQUE INDEX opt_name '(' name_list ')' opt_storing
```
Comment (tamird) [inline sql/parser/sql.y]: i'd rather we removed this variant, but that's personal preference

**Comment by tamird** [comment]
LGTM

File: sql/parser/sql.y (hunk 2 above)
Comment (petermattis) [inline sql/parser/sql.y]: Because the functionality is already supported elsewhere in the grammar? Or something else?

Comment (tamird) [inline sql/parser/sql.y]: Oh, this is shared by CREATE TABLE and ADD INDEX now, so removing it here would also disable CREATE UNIQUE INDEX. Apologies for the noise.

Comment (petermattis) [inline sql/parser/sql.y]: This rule isn't actually shared with `CREATE INDEX`. The syntax is similar, but the rule is distinct. Hope that doesn't change your mind.

Comment (tamird) [inline sql/parser/sql.y]: I guess we'll have 3 ways of creating unique indexes :thought_balloon:

Comment (petermattis) [inline sql/parser/sql.y]: I'd argue this is only syntactic sugar and that there are only two ways.

**Pull request merged**